consumer/consumererror/README.md

```diff
@@ -0,0 +1,117 @@
+- `consumererror.NewHTTP` for errors resulting from an HTTP call with an error status code.
+- `consumererror.NewGRPC` for errors resulting from a gRPC call with an error status code.
+- `consumererror.NewComponent` for metadata around the component that produced the error.
+
+Only retryable errors will be retried, all other errors are considered permanent
+and will not be retried.
+
+Errors can be joined by passing them to a call to `errors.Join`.
+
+
+### Per-signal retryable errors
+
+If an error is considered transient and data processing should be retried, the data
+that should be retried should be created using `NewTraces`, `NewMetrics` or `NewLogs`
+with the data that should be retried, e.g. `NewTraces(err, traces)`. The data that
+is contained in the error will be the only data that is retried, the caller should
+not use its copy of the data for the retry.
+
+## Other considerations
+
+To keep error analysis simple when looking at an error upstream in a pipeline,
+the component closest to the source of an error or set of errors should make a
+decision about the nature of the error. The following are a few places where
+special considerations may need to be made.
+
+### Fanouts
+
+#### Permanent errors
+
+When a fanout component receives multiple permanent errors from downstream in
+the pipeline, it should determine the error with the highest precedence and
+forward that upward in the pipeline. If a precedence cannot be determined,
+the component should forward all errors.
+
+#### Retryable errors
+
+When a fanout component receives a retryable error, it must only retry the data
+the failed pipeline branch.
+
+#### Mixed errors
+
+When a mix of retryable and permanent errors are received from downstream
+pipeline branches, the fanout component should continue all retries for
+retryable data and return a permanent error upstream regardless of whether
+any of the retries succeed. The component performing retries should capture
+which pipelines succeeded or failed using the Collector's pipeline observability
+mechanisms.
+
+### Signal conversion
+
+When converting between signals in a pipeline, it is expected that the component
+performing the conversion should perform the translation necessary in the error
+for any signal item counts. If the converted count cannot be determined, the full
+count of pre-converted signals should be returned.
+
+### Asynchronous processing
+
+The use of any components that do asynchronous processing in a pipeline will cut
+off error backpropagation at the asynchronous component. The asynchronous
+component may communicate error information using the Collector's own signals.
+
+## Examples
+
+Creating an error:
+
+```golang
+errors.Join(
+    consumererror.NewComponent(exp.componentID),
+    consumererror.NewTraces(
+        error,
+        traces,
+        consumererror.WithRetryDelay(2 * time.Duration)
+    ),
+    consumererror.NewHTTP(error, http.StatusTooManyRequests)
```

Comment:
I don't think such composition is the right approach to communicating a particular error. How is a component handling an error supposed to understand if this is one error or three errors from a fanout? I think we should keep `errors.Join` only for fanout and not to compose a particular error. 

Can we have the component, retryable/not-retryable, grps/http be fields of one error type?

Reply:
>Can we have the component, retryable/not-retryable, grps/http be fields of one error type?

This was the original approach, but I modified it in response to https://github.com/open-telemetry/opentelemetry-collector/pull/9041#discussion_r1417768965. I still agree with my initial assessment that it still feels a bit odd, though it does perhaps provide some flexibility.

>How is a component handling an error supposed to understand if this is one error or three errors from a fanout?

I still think the fanout is the only thing that can meaningfully handle this since it is the only place where we have the context of where the error came from. Even if we put all fields on a single error object, the component will still only get an list of errors with unstable ordering and will have to e.g. resend data to call components if there's a retryable error. In my opinion, if there's a fanout, we need to have the fanout component make the determination of which error to forward upstream.

Reply:
If the fanout can't determine a precedence, you can still get each individually wrapped error using the `Unwrap` method on one of the wrapped errors returned by the fanout, though maybe this isn't a strong enough guarantee.